overwatch/src/services/state.rs

```diff
@@ -144,46 +144,131 @@ impl<Settings> ServiceState for NoState<Settings> {
 ///
 /// A [`StateHandle`] watches a stream of incoming states and triggers the
 /// attached operator handling method over it.
 pub struct StateHandle<State, Operator> {
-    watcher: StateWatcher<State>,
+    watcher: StateWatcher<Option<State>>,
     operator: Operator,
+    operator_fuse_receiver: fuse::Receiver,
 }
 
+// Clone must be used carefully. It's very likely `Operator` will be a
+// `StateOperator`, which are likely behaving in as if they were singletons.
 // Clone is implemented manually because auto deriving introduces an unnecessary
 // Clone bound on T.
 impl<State, Operator> Clone for StateHandle<State, Operator>
 where
+    State: Clone,
     Operator: Clone,
 {
     fn clone(&self) -> Self {
         Self {
             watcher: self.watcher.clone(),
             operator: self.operator.clone(),
+            operator_fuse_receiver: self.operator_fuse_receiver.resubscribe(),
         }
     }
 }
 
 impl<State, Operator> StateHandle<State, Operator> {
-    pub fn new(initial_state: State, operator: Operator) -> (Self, StateUpdater<State>) {
+    pub const fn watcher(&self) -> &StateWatcher<Option<State>> {
+        &self.watcher
+    }
+
+    pub const fn operator(&self) -> &Operator {
+        &self.operator
+    }
+}
+
+impl<State, Operator> StateHandle<State, Operator>
+where
+    State: Clone,
+{
+    pub fn new(
+        operator: Operator,
+        initial_state: Option<State>,
+        operator_fuse_receiver: fuse::Receiver,
+    ) -> (Self, StateUpdater<State>) {
         let (sender, receiver) = channel(initial_state);
         let watcher = StateWatcher { receiver };
         let updater = StateUpdater {
             sender: Arc::new(sender),
         };
 
-        (Self { watcher, operator }, updater)
+        (
+            Self {
+                watcher,
+                operator,
+                operator_fuse_receiver,
+            },
+            updater,
+        )
+    }
+}
+
+impl<State, Operator> StateHandle<State, Operator>
+where
+    State: Clone + Send + Sync + 'static,
+    Operator: StateOperator<State = State>,
+{
+    /// Wait for new state updates and run the operator handling method.    
+    pub async fn run(self) {
+        let Self {
+            watcher,
+            mut operator,
+            mut operator_fuse_receiver,
+        } = self;
+
+        let mut state_stream = WatchStream::new(watcher.receiver);
+        loop {
+            tokio::select! {
+                 _ = operator_fuse_receiver.recv() => {
+                     dbg!("StateHandle's Operator loop received a fuse signal.");
+                     break;
+                 }
+                Some(state) = state_stream.next() => {
+                    dbg!("StateHandle's Stream received a state. Forwarding to Operator.");
+                    Self::process_state(&mut operator, state).await;
+                }
+            }
+        }
+
+        dbg!("Attempting to fetch the last state from StateHandle's Stream.");
```

Comment:
Could we group this into a separate function, which feels a lot like a `teardown` function?